File: app/src/main/java/id/co/pcsindonesia/ia/diagnostic/GlobalHelper.java
package id.co.pcsindonesia.ia.diagnostic;

import android.Manifest;
import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;
import android.content.pm.PackageInfo;
import android.content.pm.PackageManager;
import android.graphics.Color;
import android.net.ConnectivityManager;
import android.net.NetworkInfo;
import android.net.wifi.WifiInfo;
import android.net.wifi.WifiManager;
import android.os.BatteryManager;
import android.os.Build;
import android.os.PowerManager;
import android.telephony.CellInfoGsm;
import android.telephony.CellInfoLte;
import android.telephony.CellInfoWcdma;
import android.telephony.CellSignalStrengthGsm;
import android.telephony.CellSignalStrengthLte;
import android.telephony.CellSignalStrengthWcdma;
import android.telephony.TelephonyManager;
import android.text.SpannableString;
import android.text.Spanned;
import android.text.format.Formatter;
import android.text.style.ForegroundColorSpan;
import android.util.Log;

import androidx.core.content.ContextCompat;

import com.sunmi.pay.hardware.aidlv2.AidlConstantsV2;
import com.sunmi.pay.hardware.aidlv2.system.BasicOptV2;

import java.io.File;
import java.net.Inet4Address;
import java.net.InetAddress;
import java.net.NetworkInterface;
import java.net.SocketException;
import java.text.DecimalFormat;
import java.text.SimpleDateFormat;
import java.util.Collections;
import java.util.Enumeration;
import java.util.List;

import static android.content.Context.BATTERY_SERVICE;
import static android.content.Context.WIFI_SERVICE;


public class GlobalHelper {

    public static final String BRILINK_PACKAGE = "id.co.bri.brilinkmobile";
    public static final String DECO_PACKAGE = "id.co.pcsindonesia.ia.dcm";
    public static final String APPSTORE_PACKAGE = "woyou.market";

    public static final SimpleDateFormat ddMMyyyy = new SimpleDateFormat("dd-MM-yyyy");

    public static final int TYPE_WIFI = 1;
    public static final int TYPE_MOBILE = 2;
    public static final int TYPE_NOT_CONNECTED = 0;
    public static final String NETWORK_STATUS_NOT_CONNECTED = "Not Connected";
    public static final String NETWORK_STATUS_WIFI = "Wifi";
    public static final String NETWORK_STATUS_MOBILE = "Mobile";



    public static String getAppVersion(Context context){
        String version = "";
        try {
            PackageInfo pInfo = context.getPackageManager().getPackageInfo(context.getPackageName(), 0);
            version = pInfo.versionName;
        } catch (PackageManager.NameNotFoundException e) {
            e.printStackTrace();
        }
        return version;
    }

    public static String getAppVersionByPackege(Context context, String packageName){
        String version = "";
        try {
            PackageInfo pInfo = context.getPackageManager().getPackageInfo(packageName, 0);
            version = pInfo.versionName;
        } catch (PackageManager.NameNotFoundException e) {
            e.printStackTrace();
        }
        return version;
    }

    public static String getDeviceModel(){
        return Build.MODEL;
    }

    public static String getAppPackageName(Context context){
        return context.getPackageName();
    }

    public static String getIPAddress() {
        try {
            for (Enumeration<NetworkInterface> en = NetworkInterface.getNetworkInterfaces(); en.hasMoreElements();) {
                NetworkInterface intf = en.nextElement();
                for (Enumeration<InetAddress> enumIpAddr = intf.getInetAddresses(); enumIpAddr.hasMoreElements();) {
                    InetAddress inetAddress = enumIpAddr.nextElement();
                    if (!inetAddress.isLoopbackAddress() && inetAddress instanceof Inet4Address) {
                        return inetAddress.getHostAddress();
                    }
                }
            }
        } catch (SocketException ex) {
            ex.printStackTrace();
        }
        return null;
    }

    public static String getSN(){
        String sn = "";
        if(App.isSunmiConnected) {
            BasicOptV2 basicOptV2 = App.mBasicOptV2;
            try {
                sn = basicOptV2.getSysParam(AidlConstantsV2.SysParam.SN);
            } catch (Exception e) {
                sn = "";
            }
        }
        return sn;
    }

    public static String getAppsVersion(Context context, String packageName){
        String version = "-";
        try {
            PackageInfo pInfo = context.getPackageManager().getPackageInfo(packageName, 0);
            version = pInfo.versionName;
        } catch (PackageManager.NameNotFoundException e) {
            e.printStackTrace();

        }
        return version;
    }

    public static long getAppsVersionCode(Context context, String packageName){
        long version = 0;
        String temp = "";
        try {
            PackageInfo pInfo = context.getPackageManager().getPackageInfo(packageName, 0);
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.P) {
                temp = pInfo.versionName; // avoid huge version numbers and you will be ok
            } else {
                //noinspection deprecation
                temp = pInfo.versionName;
            }
        } catch (PackageManager.NameNotFoundException e) {
            e.printStackTrace();
        }

        version = Long.parseLong(temp.replace(".", ""));

        return version;
    }

    public static int getWifiStatus(Context context) {
        WifiManager wimanager = (WifiManager) context.getSystemService(Context.WIFI_SERVICE);
        boolean macAddress = wimanager.isWifiEnabled();
        if (macAddress) {
            return 1;
        }else{
            return 0;
        }
    }

    public static String getMacAddress(Context context) {
        WifiManager wimanager = (WifiManager) context.getSystemService(Context.WIFI_SERVICE);
        String macAddress = wimanager.getConnectionInfo().getMacAddress();
        if (macAddress == null) {
            macAddress = "";
        }
        return macAddress;
    }

    public static int getWifiSignal(Context context){
        int signal = -100;
        WifiManager wimanager = (WifiManager) context.getSystemService(Context.WIFI_SERVICE);
        signal = wimanager.getConnectionInfo().getRssi();
        return signal;
    }

    public static int getSimSignal(Context context){
        int signal = -127;
        if (ContextCompat.checkSelfPermission( context, Manifest.permission.READ_PHONE_STATE ) == PackageManager.PERMISSION_GRANTED ) {
            TelephonyManager tm = (TelephonyManager) context.getSystemService(Context.TELEPHONY_SERVICE);

            String signals = "";
            try{
                CellInfoGsm cellInfoGsm = (CellInfoGsm) tm.getAllCellInfo().get(0);
                CellSignalStrengthGsm cellSignalStrengthGsm = cellInfoGsm.getCellSignalStrength();
                signals += String.valueOf(cellSignalStrengthGsm.getDbm());
            }catch (RuntimeException e){
                e.printStackTrace();
            }
            try{
                CellInfoWcdma cellInfoGsm = (CellInfoWcdma) tm.getAllCellInfo().get(0);
                CellSignalStrengthWcdma cellSignalStrengthGsm = cellInfoGsm.getCellSignalStrength();
                signals += String.valueOf(cellSignalStrengthGsm.getDbm());
            }catch (RuntimeException e){
                e.printStackTrace();
            }
            try{
                CellInfoLte cellInfoGsm = (CellInfoLte) tm.getAllCellInfo().get(0);
                CellSignalStrengthLte cellSignalStrengthGsm = cellInfoGsm.getCellSignalStrength();
                signals += String.valueOf(cellSignalStrengthGsm.getDbm());
            }catch (RuntimeException e){
                e.printStackTrace();
            }
            try {
                signal = Integer.parseInt(signals);
            }catch (Exception e){
                e.printStackTrace();
            }
            return signal;
        }
        return -127;
    }

    @SuppressWarnings("deprecation")
    public String getIPAddress(Context context) {
        if(getConnectivityStatus(context) == id.co.pcsindonesia.ia.diagnostic.GlobalHelper.TYPE_MOBILE) {
            try {
                List<NetworkInterface> interfaces = Collections.list(NetworkInterface.getNetworkInterfaces());
                for (NetworkInterface intf : interfaces) {
                    List<InetAddress> addrs = Collections.list(intf.getInetAddresses());
                    for (InetAddress addr : addrs) {
                        if (!addr.isLoopbackAddress()) {
                            return addr.getHostAddress();
                        }
                    }
                }
            } catch (Exception ex) {
            } // for now eat exceptions
            return "";
        }else{
            WifiManager wifiMgr = (WifiManager) context.getApplicationContext().getSystemService(WIFI_SERVICE);
            WifiInfo wifiInfo = wifiMgr.getConnectionInfo();
            return  Formatter.formatIpAddress(wifiInfo.getIpAddress());
        }
    }

    public static String getIMEI(Context context){
        if (ContextCompat.checkSelfPermission( context, Manifest.permission.READ_PHONE_STATE ) == PackageManager.PERMISSION_GRANTED ) {
            TelephonyManager tm = (TelephonyManager) context.getSystemService(Context.TELEPHONY_SERVICE);
            return tm.getDeviceId();
        }
        return null;
    }

    public static String getICCID(Context context){
        String result = "-";
        if (ContextCompat.checkSelfPermission( context, Manifest.permission.READ_PHONE_STATE ) == PackageManager.PERMISSION_GRANTED ) {
            TelephonyManager tm = (TelephonyManager) context.getSystemService(Context.TELEPHONY_SERVICE);
            result = tm.getSimSerialNumber();
        }
        if(result == null) result = "-";
        return result;
    }

    public static String getPercentBattery(Context context){
        float batLevel = 0;
        BatteryManager bm = (BatteryManager)context.getSystemService(BATTERY_SERVICE);

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
            batLevel = bm.getIntProperty(BatteryManager.BATTERY_PROPERTY_CAPACITY);
        }
        if(batLevel == 0){
            Intent batteryIntent = context.registerReceiver(null, new IntentFilter(Intent.ACTION_BATTERY_CHANGED));
            int level = batteryIntent.getIntExtra(BatteryManager.EXTRA_LEVEL, -1);
            int scale = batteryIntent.getIntExtra(BatteryManager.EXTRA_SCALE, -1);

            // Error checking that probably isn't needed but I added just in case.
            if(level == -1 || scale == -1) {
                batLevel =  50.0f;
            }

            batLevel =  ((float)level / (float)scale) * 100.0f;
        }
        return batLevel+"";
    }

    public static void turnScreenOn(Context context){
        PowerManager manager = (PowerManager) context.getSystemService(Context.POWER_SERVICE);
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
            if (!manager.isInteractive()) {
                Log.e("ThreadService", "Wakelock: " + manager.isInteractive());
                PowerManager.WakeLock fullWakeLock = manager.newWakeLock((PowerManager.FULL_WAKE_LOCK | PowerManager.SCREEN_BRIGHT_WAKE_LOCK | PowerManager.ACQUIRE_CAUSES_WAKEUP), "myapp:mywakelocktag");
                fullWakeLock.acquire(); // turn on
                fullWakeLock.release();
            }
        }

    }

    public static int getConnectivityStatus(Context context) {
        ConnectivityManager cm = (ConnectivityManager) context.getSystemService(Context.CONNECTIVITY_SERVICE);

        if (ContextCompat.checkSelfPermission( context, Manifest.permission.ACCESS_NETWORK_STATE ) == PackageManager.PERMISSION_GRANTED ) {
            NetworkInfo activeNetwork = cm.getActiveNetworkInfo();
            if (null != activeNetwork) {
                if (activeNetwork.getType() == ConnectivityManager.TYPE_WIFI)
                    return TYPE_WIFI;

                if (activeNetwork.getType() == ConnectivityManager.TYPE_MOBILE)
                    return TYPE_MOBILE;
            }
        }
        return TYPE_NOT_CONNECTED;
    }

    public static String getConnectivityStatusString(Context context) {
        int conn = getConnectivityStatus(context);
        String status = "";
        if (conn == TYPE_WIFI) {
            status = NETWORK_STATUS_WIFI;
        } else if (conn == TYPE_MOBILE) {
            status = NETWORK_STATUS_MOBILE;
        } else if (conn == TYPE_NOT_CONNECTED) {
            status = NETWORK_STATUS_NOT_CONNECTED;
        }
        return status;
    }

    public static SpannableString fileSizeShorter(String fileSize, boolean withSuffix){
        DecimalFormat decimalFormat = new DecimalFormat("#,###.#");
        Double fileSizeLong = Double.parseDouble(fileSize);
        Double finalSize=0.0;
        String suffix = "B";
        if(fileSizeLong >= 1024){
            if(fileSizeLong>= 1024*1024){
                if(fileSizeLong >=1024*1024*1024){
                    finalSize = fileSizeLong / (1024*1024*1024);
                    suffix = "GB";
                }else{
                    finalSize = fileSizeLong / (1024*1024);
                    suffix = "MB";
                }
            }else{
                finalSize = fileSizeLong / 1024;
                suffix = "KB";
            }
        }else{
            finalSize = fileSizeLong;
            suffix = "B";
        }
        if(withSuffix){
            SpannableString string = new SpannableString(decimalFormat.format(finalSize) + suffix);
            switch (suffix){
                case "GB":
                    string.setSpan(new ForegroundColorSpan(Color.parseColor("#FF0000")),string.toString().indexOf(suffix),
                            string.toString().indexOf(suffix)+suffix.length(), Spanned.SPAN_INCLUSIVE_INCLUSIVE);
                    break;
                case "MB":
                    string.setSpan(new ForegroundColorSpan(Color.parseColor("#FFA500")),string.toString().indexOf(suffix),
                            string.toString().indexOf(suffix)+suffix.length(), Spanned.SPAN_INCLUSIVE_INCLUSIVE);
                    break;
                case "KB":
                    string.setSpan(new ForegroundColorSpan(Color.parseColor("#008000")),string.toString().indexOf(suffix),
                            string.toString().indexOf(suffix)+suffix.length(), Spanned.SPAN_INCLUSIVE_INCLUSIVE);
                    break;
                case "B":
                    string.setSpan(new ForegroundColorSpan(Color.parseColor("#0000FF")),string.toString().indexOf(suffix),
                            string.toString().indexOf(suffix)+suffix.length(), Spanned.SPAN_INCLUSIVE_INCLUSIVE);
                    break;

            }
            return string;
        }else{
            SpannableString string = new SpannableString(decimalFormat.format(finalSize));
            return string;
        }

    }

    public static boolean isPackageInstalled(String packageName, Context context) {
        boolean cek;
        PackageManager manager = context.getPackageManager();
        try {
            manager.getApplicationInfo(packageName, 0);
            cek = true;
        } catch (PackageManager.NameNotFoundException e) {
            cek = false;
        }
        return cek;
    }

    public static void deleteRecursive(String dir_path){
        File dir = new File(dir_path);
        if (dir.isDirectory())
        {
            String[] children = dir.list();
            for (int i = 0; i < children.length; i++)
            {
                new File(dir, children[i]).delete();
            }
            Log.e("GlobalHelper", "deleteRecursive; "+ children.length+" ; "+dir_path);
        }
    }

}

File: app/src/main/java/id/co/pcsindonesia/ia/diagnostic/launcher/GlobalHelper.kt
package id.co.pcsindonesia.pcslauncher

import android.Manifest
import android.annotation.SuppressLint
import android.app.Activity
import android.content.Context
import android.content.Intent
import android.content.IntentFilter
import android.content.pm.ApplicationInfo
import android.content.pm.PackageManager
import android.net.wifi.WifiManager
import android.os.BatteryManager
import android.os.Build
import android.telephony.CellInfoGsm
import android.telephony.CellInfoLte
import android.telephony.CellInfoWcdma
import android.telephony.TelephonyManager
import android.view.View
import android.view.inputmethod.InputMethodManager
import androidx.core.content.ContextCompat
import com.google.gson.Gson
import com.sunmi.pay.hardware.aidl.AidlConstants
import id.co.pcsindonesia.ia.diagnostic.launcher.App.Companion.isSunmiConnected
import id.co.pcsindonesia.ia.diagnostic.launcher.App.Companion.mBasicOptV2
import id.co.pcsindonesia.ia.diagnostic.launcher.LogUtil
import id.co.pcsindonesia.ia.diagnostic.printer.PrinterManagement
import io.jsonwebtoken.JwtException
import io.jsonwebtoken.Jwts
import io.jsonwebtoken.security.Keys
import org.json.JSONArray
import org.json.JSONException
import org.json.JSONObject
import java.io.File
import java.io.FileWriter
import java.io.IOException
import java.net.Inet4Address
import java.net.NetworkInterface
import java.net.SocketException
import java.text.SimpleDateFormat
import java.util.*

class GlobalHelper {

    companion object {
        var TAG = "GlobalHelper"

        val BRILINK = 1

        @SuppressLint("SimpleDateFormat")
        val ddMMyyyy = SimpleDateFormat("dd-MM-yyyy")

        @SuppressLint("SimpleDateFormat")
        val ymdh = SimpleDateFormat("yyyy-MM-dd HH")

        @JvmField
        var SUNMI_BASE_MODEL: String = "P1_4G"

        @JvmField
        var SUNMI_BASE_MODEL_P2: String = "P2"

        val JWT_KEY = "euOJUsxWZ1oa1uq1dvSYLMhRwDmwONasbyfUfKmR8WQlb5hwUWrYBUw8mjaRRG8R"

        fun getDeviceModel(): String? {
            return Build.MODEL
        }

        fun getTimeHBPrimary(): Int {
            val res = 300 //default (detik)`
            return res
        }

        fun getTimeHBSecondary(): Int {
            val res = 3600 //default (detik)`
            return res
        }



        fun getToken(data : JSONObject) : String {
            val claim = data.toMap()
            val mapHeader = mapOf<String,String>("typ" to "JWT")
            val key = Keys.hmacShaKeyFor(JWT_KEY.toByteArray())
            val jws = Jwts.builder()
                .setExpiration(getExpirationDate())
                .setIssuer("launcher.pcsindoensia.co.id")
//                .setAudience(ConstantsShared.jwtAudience)
//                .setSubject(ConstantsShared.jwtSubject)
//                .setNotBefore(Date())
                .setIssuedAt(Date())
                .setHeader(mapHeader)
                .signWith(key)
                .claim("msg", claim)
                .compact();
            return jws
        }

        private fun getExpirationDate(): Date? {
            val cal = Calendar.getInstance(TimeZone.getTimeZone("GMT+7"))
            cal.time = Date()
            cal.add(Calendar.MINUTE, 10)

            LogUtil.e("DATE",cal.time.toString())

            return cal.time

        }

        fun parseToken(data: String): JSONObject? {
            return try {
                val key = Keys.hmacShaKeyFor(JWT_KEY.toByteArray())
                val jws = Jwts.parserBuilder()
                    .setSigningKey(key)
                    .build()
                    .parseClaimsJws(data)
                    .body["msg"]

                JSONObject(Gson().toJson(jws,LinkedHashMap::class.java))
            } catch (e : JwtException) {
                e.printStackTrace()
                LogUtil.e(TAG,e.message)
                JSONObject()
            }
        }

        fun getSN(): String? {
            var sn = ""
            if (getDeviceModel()!!.startsWith(SUNMI_BASE_MODEL) || getDeviceModel()!!.startsWith(
                    SUNMI_BASE_MODEL_P2
                )
            ) {
                if (isSunmiConnected) {
                    val basicOptV2 = mBasicOptV2
                    sn = try {
                        basicOptV2!!.getSysParam(AidlConstants.SysParam.SN)
                    } catch (e: java.lang.Exception) {
                        LogUtil.e(TAG, "getSN-Error: " + e.message)
                        ""
                    }
                }
            }
            return sn
        }

        fun hideKeyboard(activity: Activity) {
            val imm = activity.getSystemService(
                Activity.INPUT_METHOD_SERVICE
            ) as InputMethodManager
            var view = activity.currentFocus
            if (view == null) {
                view = View(activity)
            }
            imm.hideSoftInputFromWindow(view.windowToken, 0)
        }

        fun openKeyboard(activity: Activity) {
            val imm =
                activity.getSystemService(Context.INPUT_METHOD_SERVICE) as InputMethodManager
            imm.toggleSoftInput(InputMethodManager.SHOW_FORCED, 0)
        }
        fun getBatteryBundle(context: Context): JSONObject? {
            var batteryBundle = JSONObject()
            var batLevel = 0f
            val bm = context.getSystemService(Context.BATTERY_SERVICE) as BatteryManager
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
                batLevel = bm.getIntProperty(BatteryManager.BATTERY_PROPERTY_CAPACITY).toFloat()
            }
            if (batLevel == 0f) {
                val batteryIntent =
                    context.registerReceiver(null, IntentFilter(Intent.ACTION_BATTERY_CHANGED))
                val level = batteryIntent!!.getIntExtra(BatteryManager.EXTRA_LEVEL, -1)
                val scale = batteryIntent.getIntExtra(BatteryManager.EXTRA_SCALE, -1)

                // Error checking that probably isn't needed but I added just in case.
                if (level == -1 || scale == -1) {
                    batLevel = 50.0f
                }
                batLevel = level.toFloat() / scale.toFloat() * 100.0f
            }



            var batTemp = 0f
            val intent = context.registerReceiver(null, IntentFilter(Intent.ACTION_BATTERY_CHANGED))

            batTemp = intent!!.getIntExtra(BatteryManager.EXTRA_TEMPERATURE, 0).toFloat() / 10


            var batHealth = when (intent!!.getIntExtra(BatteryManager.EXTRA_HEALTH, 0)){
                BatteryManager.BATTERY_HEALTH_COLD -> "Cold"
                BatteryManager.BATTERY_HEALTH_DEAD -> "Dead"
                BatteryManager.BATTERY_HEALTH_GOOD -> "Good"
                BatteryManager.BATTERY_HEALTH_OVERHEAT -> "Overheat"
                BatteryManager.BATTERY_HEALTH_OVER_VOLTAGE -> "Overvoltage"
                BatteryManager.BATTERY_STATUS_UNKNOWN -> "Unknown"
                BatteryManager.BATTERY_HEALTH_UNSPECIFIED_FAILURE -> "Unspecified"
                else -> "Error"
            }




            batteryBundle.put("level", batLevel.toString())
            batteryBundle.put("temperature", batTemp.toString())
            batteryBundle.put("health", batHealth)

            return batteryBundle
        }

        fun getListApp (context: Context) : JSONObject? {
            val pManager = context.packageManager
            val i = Intent(Intent.ACTION_MAIN,null).addCategory(Intent.CATEGORY_LAUNCHER)
            val apps = pManager.getInstalledPackages(0)
            var listApp = JSONObject()


            for (app in apps){
                var ai = pManager.getApplicationInfo(app.packageName,0)
                if (ai.flags and ApplicationInfo.FLAG_SYSTEM === 0) {
                    var aa = JSONObject()
                    aa.put("version_name",app.versionName)
                    aa.put("version_code",app.versionCode)
                    aa.put("package_name",app.packageName)
                    listApp.put(app.applicationInfo.loadLabel(pManager).toString(),aa)
                }

            }
            return  listApp
        }

        fun getLocation (context: Context) : JSONObject? {
            val sharedPreferences = SharedPreferencesHelper(context)
            sharedPreferences.init()
            var location = JSONObject()
            var longitude: String? = "0"
            var latitude: String? = "0"

            if (sharedPreferences.getLocation() != null) {
                val strLoc = sharedPreferences.getLocation()
                try {
                    val obj = JSONObject(strLoc)
                    longitude = obj.getString("longitude")
                    latitude = obj.getString("latitude")
                } catch (e: JSONException) {
                    e.printStackTrace()
                    LogUtil.e(TAG, "postPrimaryHb-Location convert error: " + e.message)
                }
            }
            location.put("longitude",longitude)
            location.put("latitude",latitude)

            return location
        }

        fun getSignal(context: Context) : JSONObject? {
            var signal = JSONObject()

            var wsignal = -100
            if (ContextCompat.checkSelfPermission(
                    context,
                    Manifest.permission.ACCESS_WIFI_STATE
                ) == PackageManager.PERMISSION_GRANTED
            ) {
                val wimanager = context.getSystemService(Context.WIFI_SERVICE) as WifiManager
                wsignal = wimanager.connectionInfo.rssi
            } else {
                LogUtil.e(TAG, "WIFI STATE NOT GRANTED")
            }

            var ssignal = -127
            if (ContextCompat.checkSelfPermission(
                    context,
                    Manifest.permission.READ_PHONE_STATE
                ) == PackageManager.PERMISSION_GRANTED
            ) {
                val tm = context.getSystemService(Context.TELEPHONY_SERVICE) as TelephonyManager
                var signals = ""
                try {
                    if (!tm.allCellInfo.isEmpty()) {
                        val cellInfoWcdma : CellInfoWcdma = tm.allCellInfo[0] as CellInfoWcdma
                        val cellSignalStrengthWcdma = cellInfoWcdma.cellSignalStrength
                        signals += cellSignalStrengthWcdma.dbm.toString()
                    }
                } catch (e: RuntimeException) {
                    e.printStackTrace()
                    LogUtil.e(TAG, "getSimSignal error : $e")
                    try {
                        if (!tm.allCellInfo.isEmpty()) {
                            val cellInfoGsm = tm.allCellInfo[0] as CellInfoGsm
                            val cellSignalStrengthGsm = cellInfoGsm.cellSignalStrength
                            signals += cellSignalStrengthGsm.dbm.toString()
                        }

                    } catch (e: RuntimeException) {
                        e.printStackTrace()
                        LogUtil.e(TAG, "getSimSignal error : $e")
                        try {
                            if (!tm.allCellInfo.isEmpty()) {
                                val cellInfoLte = tm.allCellInfo[0] as CellInfoLte
                                val cellSignalStrengthLte = cellInfoLte.cellSignalStrength
                                signals += cellSignalStrengthLte.dbm.toString()
                            }
                        } catch (e: RuntimeException) {
                            e.printStackTrace()
                            LogUtil.e(TAG, "getSimSignal error : $e")
                        }
                    }
                }

                try {
                    ssignal = signals.toInt()
                } catch (e: java.lang.Exception) {
                    e.printStackTrace()
                    LogUtil.e(TAG, "getSimSignal error : $e")
                }
            }

            signal.put("wifi",wsignal)
            signal.put("sim",ssignal)


            return signal
        }

        fun getPrinterCondition(context: Context) : JSONObject {
            var printer = JSONObject()
            var condition = ""
            var pm = PrinterManagement(context)

            var res = pm.checkCondition()
            var version = pm.checkVersion()

            when (res){
                1 -> condition = "Normal"
                2 -> condition = "Abnormal (2)"
                3 -> condition = "Abnormal (3)"
                4 -> condition = "Thermal paper kosong"
                5 -> condition = "Overheat"
                6 -> condition = "Cover terbuka"
                7 -> condition = "Cutter error"
                8 -> condition = "Cutter recovered"
                9 -> condition = "Black mark not detected"
                505 -> condition = "Printer not detected"
                507 -> condition = "Printer firmware update failed"
                else -> "null"
            }

            printer.put("version",version)
            printer.put("condition",condition)

            return printer
        }

        fun getDevice() : JSONObject? {
            var device = JSONObject()
            device.put("model", Build.MODEL)
            device.put("manufacturer", Build.MANUFACTURER)
            device.put("brand",Build.BRAND)

            return device
        }

        fun getIMEI(context: Context): String? {
            var imei = "-"
            if (ContextCompat.checkSelfPermission(
                    context,
                    Manifest.permission.READ_PHONE_STATE
                ) == PackageManager.PERMISSION_GRANTED
            ) {
                val tm = context.getSystemService(Context.TELEPHONY_SERVICE) as TelephonyManager
                imei = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
                    tm.imei
                } else {
                    tm.deviceId
                }
            }
            return imei
        } fun getMacAddress(context: Context): String? {
            var macAddress = ""
            if (ContextCompat.checkSelfPermission(
                    context,
                    Manifest.permission.ACCESS_WIFI_STATE
                ) == PackageManager.PERMISSION_GRANTED
            ) {

                val wimanager = context.getSystemService(Context.WIFI_SERVICE) as WifiManager
                macAddress = wimanager.connectionInfo.macAddress
                if (macAddress == null) {
                    macAddress = ""
                }
            } else {
                LogUtil.e(TAG, "WIFI STATE NOT GRANTED")
            }
            return macAddress
        }

        fun getIPAddress(): String? {
            try {
                val en = NetworkInterface.getNetworkInterfaces()
                while (en.hasMoreElements()) {
                    val intf = en.nextElement()
                    val enumIpAddr = intf.inetAddresses
                    while (enumIpAddr.hasMoreElements()) {
                        val inetAddress = enumIpAddr.nextElement()
                        if (!inetAddress.isLoopbackAddress && inetAddress is Inet4Address) {
                            return inetAddress.getHostAddress()
                        }
                    }
                }
            } catch (ex: SocketException) {
                ex.printStackTrace()
            }
            return null
        }


        fun updateFiles(directory: String?, filename: String?, sBody: String?) {
            try {
                val file = File(directory, filename)
                val root = File(directory)
                if (!root.exists()) {
                    root.mkdirs()
                }
                if (file.exists()) {
                    file.delete()
                }
                val gpxfile = File(root, filename)
                val writer = FileWriter(gpxfile)
                writer.write(sBody)
                writer.flush()
                writer.close()
            } catch (e: IOException) {
                e.printStackTrace()
            }
        }

        fun getCacheDirHeartbeat(context: Context): String {
            val baseDir = context.cacheDir
            val path = "$baseDir/pcs_launcher/data/data/heartbeat"
            val file = File(path)
            if (!file.exists()) file.mkdirs()
            return file.absolutePath + "/"
        }

        fun getResponseFilename(): String {
            return "last_heartbeat.txt"
        }


        fun fullScreencall(activity: Activity) {
            if (Build.VERSION.SDK_INT > 11 && Build.VERSION.SDK_INT < 19) { // lower api
                val v: View = activity.window.decorView
                v.setSystemUiVisibility(View.GONE)
            } else if (Build.VERSION.SDK_INT >= 19) {
                //for new api versions.
                val decorView: View = activity.window.decorView
                val uiOptions: Int =
                    View.SYSTEM_UI_FLAG_HIDE_NAVIGATION or View.SYSTEM_UI_FLAG_IMMERSIVE_STICKY
                decorView.setSystemUiVisibility(uiOptions)
            }
        }

        fun JSONObject.toMap(): Map<String, *> = keys().asSequence().associateWith {
            when (val value = this[it])
            {
                is JSONArray ->
                {
                    val map = (0 until value.length()).associate { Pair(it.toString(), value[it]) }
                    JSONObject(map).toMap().values.toList()
                }
                is JSONObject -> value.toMap()
                JSONObject.NULL -> null
                else            -> value
            }
        }

    }

}


